File: backend/src/main/java/com/plantpal/plant/service/PlantService.java
package com.plantpal.plant.service;

import com.plantpal.plant.dto.CreatePlantRequest;
import com.plantpal.plant.dto.PlantResponse;
import com.plantpal.plant.dto.SaveIdentificationAsPlantRequest;
import com.plantpal.plant.dto.UpdatePlantRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface PlantService {

  PlantResponse createPlant(CreatePlantRequest request, Long userId);

  PlantResponse updatePlant(Long id, UpdatePlantRequest request, Long userId);

  void archivePlant(Long id, Long userId);

  Page<PlantResponse> getUserPlants(Long userId, Pageable pageable);

  PlantResponse getPlant(Long id, Long userId);

  PlantResponse saveFromIdentification(SaveIdentificationAsPlantRequest request, Long userId);
}

File: frontend/src/app/features/plant/services/plant.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../../environments/environment';
import { ApiResponse, PageResponse } from '@plantpal/shared-core';
import { CreatePlantRequest, PlantResponse, SaveIdentificationAsPlantRequest, UpdatePlantRequest } from '../models/plant.model';

@Injectable()
export class PlantService {
  private readonly baseUrl = `${environment.apiUrl}/plants`;

  constructor(private readonly http: HttpClient) {}

  getPlants(
    page = 0,
    size = 20,
    speciesId?: number,
  ): Observable<ApiResponse<PageResponse<PlantResponse>>> {
    let params = new HttpParams()
      .set('page', page)
      .set('size', size)
      .set('sort', 'createdAt,desc');
    if (speciesId != null) {
      params = params.set('speciesId', speciesId);
    }
    return this.http.get<ApiResponse<PageResponse<PlantResponse>>>(this.baseUrl, { params });
  }

  getPlant(id: number): Observable<ApiResponse<PlantResponse>> {
    return this.http.get<ApiResponse<PlantResponse>>(`${this.baseUrl}/${id}`);
  }

  createPlant(request: CreatePlantRequest): Observable<ApiResponse<PlantResponse>> {
    return this.http.post<ApiResponse<PlantResponse>>(this.baseUrl, request);
  }

  updatePlant(id: number, request: UpdatePlantRequest): Observable<ApiResponse<PlantResponse>> {
    return this.http.put<ApiResponse<PlantResponse>>(`${this.baseUrl}/${id}`, request);
  }

  archivePlant(id: number): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.baseUrl}/${id}`);
  }

  saveFromIdentification(request: SaveIdentificationAsPlantRequest): Observable<ApiResponse<PlantResponse>> {
    return this.http.post<ApiResponse<PlantResponse>>(`${this.baseUrl}/from-identification`, request);
  }
}
